lib/http/fetch.js

```diff
@@ -0,0 +1,159 @@
+  typeof WebStreamsPolyfill !== 'undefined' ? WebStreamsPolyfill.TransformStream :
+  undefined;
+require('../http');
+
+/**
+ * @api private
+ */
+AWS.FetchClient = AWS.util.inherit({
+  handleRequest: function handleRequest(httpRequest, httpOptions, callback, errCallback) {
+    var self = this;
+    var endpoint = httpRequest.endpoint;
+    if (!httpOptions) httpOptions = {};
+
+    var emitter = new EventEmitter();
+
+    // Construct href
+    var href = endpoint.protocol + '//' + endpoint.hostname;
+    if (endpoint.port !== 80 && endpoint.port !== 443) {
+      href += ':' + endpoint.port;
+    }
+    href += httpRequest.path;
+
+    // Construct headers
+    var headers = new Headers();
+    AWS.util.each(httpRequest.headers, function (key, value) {
+      // See: https://fetch.spec.whatwg.org/#forbidden-header-name
+      if (key !== 'Content-Length' && key !== 'Expect' && key !== 'Host') {
+        headers.set(key, value);
+      }
+    });
+
+    // Send cookies?
+    var credentials = 'omit';
+    if (httpOptions.xhrWithCredentials) {
+      credentials = 'include';
+    }
+
+    // Construct request
+    // See: https://developer.mozilla.org/en-US/docs/Web/API/Request/Request
+    var req = {
+      method: httpRequest.method,
+      headers: headers,
+      credentials: credentials
+    };
+    if (httpRequest.body) req.body = httpRequest.body;
+    var request = new Request(href, req);
+
+    // Send request
+    fetch(request).then(function(response) {
+        if (!response.ok) throw Error(response.statusText);
+        return response;
+    }).then(function(response) {
+      emitter.statusCode = response.status;
+      emitter.headers = self.parseHeaders(response.headers);
+      emitter.abort = function() { response.body.cancel(); };
+
+      httpRequest.responseStream = self.writeBody(emitter, response); // attach responseStream to httpRequest
+      httpRequest.stream = emitter; // attach emitter to httpRequest
+      emitter.didCallback = false; // Required for httpDone event
+
+      emitter.on('error', function() {
+        if (emitter.didCallback) return; emitter.didCallback = true;
+        errCallback.apply(emitter, arguments);
+      });
+
+      if (emitter.didCallback) return;
+      callback(emitter);
+      emitter.emit(
+        'headers',
+        emitter.statusCode,
+        emitter.headers,
+        response.statusText
+      );
+    }, function(err) {
+        if (emitter.didCallback) return; emitter.didCallback = true;
+        errCallback(AWS.util.error(new Error('Network Failure'), {
+        code: 'NetworkingError'
+      }));
+    }).catch(function(err) {
+      if (emitter.didCallback) return; emitter.didCallback = true;
+      errCallback(AWS.util.error(err, {
+        code: 'BadReplyError'
+      }));
+    });
+
+    return emitter;
+  },
+
+  writeBody: function writeBody(emitter, response) {
+    var body = response.body;
+    var totalBytes = parseInt(response.headers.get('Content-Length'), 10);
```

Comment:
just a heads up: if the response is encoded with gzip, br, deflate or anything then that will be what the content-length will also be (not the actual file size) so when you read the all the chunks then you will only increase the loadedBytes with uncompressed data.

which means loadedBytes might be higher than content-length in some scenarios

Reply:
This issue is very tricky indeed due to the fetch implementation, and I don't have a solution yet.

See discussion here:
https://github.com/w3c/ServiceWorker/issues/339#issuecomment-372304884

Reply:
Any suggestions are welcomed. :smiling_face_with_three_hearts:

Reply:
There is also this observer issue: https://github.com/whatwg/fetch/issues/607 but nothing have happened on that front for a long time now.

You could read the content-encoding as well and if it's null or "identity" then the content-length is the equivalent to the actual file size and it can be ok to use the progress observer. it's also possible to send a `accept-encoding: identity` request header to ask for raw bytes.  but i might best be done with a HEAD request
I don't know exactly how the aws sdk/api looks like, but there looks to be a size metadata when you fetch the file list that may be accurate.

I would probably just ditch the hole progress stuff and leave that up to the developer to handle b/c it can be unreliable. or give them the option to choose between XHR and Fetch (at least until https://github.com/whatwg/fetch/issues/607 have been implemented)

Also if you use the size option in streamsaver than you don't need to be subscribed to any progress event listener since browser native UI will show one for you. but it needs to be the uncompressed size as well - can't be the same content-length